oak_restricted_kernel/src/mm/encrypted_mapper.rs

```diff
@@ -0,0 +1,632 @@
+        }
+        if value.contains(PageTableFlags::DIRTY) {
+            flags |= BasePageTableFlags::DIRTY
+        }
+        if value.contains(PageTableFlags::HUGE_PAGE) {
+            flags |= BasePageTableFlags::HUGE_PAGE
+        }
+        if value.contains(PageTableFlags::GLOBAL) {
+            flags |= BasePageTableFlags::GLOBAL
+        }
+        // There is no equivalent of ENCRYPTED in BasePageTableFlags.
+        if value.contains(PageTableFlags::NO_EXECUTE) {
+            flags |= BasePageTableFlags::NO_EXECUTE
+        }
+        flags
+    }
+}
+
+/// Helper for mapper that assumes all memory is mapped to a given fixed offset.
+///
+/// See <x86_64::structures::paging::mapper::PageTableFrameMapping> for more details.
+pub struct PhysOffset {
+    offset: VirtAddr,
+    encryption: MemoryEncryption,
+}
+
+unsafe impl PageTableFrameMapping for PhysOffset {
+    fn frame_to_pointer(&self, frame: PhysFrame) -> *mut PageTable {
+        let virt = self.offset
+            + match self.encryption {
+                // Mapping without encryption: just do the calculation using the offset.
+                MemoryEncryption::NoEncryption => frame.start_address().as_u64(),
+                // Mapping under encryption: strip off the encrypted bit before adding the offset.
+                MemoryEncryption::Encrypted(c) => frame.start_address().as_u64() & !(1 << c),
+            };
+        virt.as_mut_ptr()
+    }
+}
+
+/// Wrapper around `FrameAllocator` that sets the encrypted bit on the allocated frame, if needed.
+///
+/// This is only meant to be used inside `EncryptedPageTable` to lie to `MappedPageTable` about the
+/// physical addresses.
+struct EncryptedFrameAllocator<'a, S: PageSize, A: FrameAllocator<S>> {
+    inner: &'a mut A,
+    encryption: MemoryEncryption,
+    phantom: PhantomData<S>,
+}
+
+impl<'a, S: PageSize, A: FrameAllocator<S>> EncryptedFrameAllocator<'a, S, A> {
+    fn new(inner: &'a mut A, encryption: MemoryEncryption) -> Self {
+        Self {
+            inner,
+            encryption,
+            phantom: PhantomData,
+        }
+    }
+}
+
+unsafe impl<'a, S: PageSize, A: FrameAllocator<S>> FrameAllocator<S>
+    for EncryptedFrameAllocator<'a, S, A>
+{
+    fn allocate_frame(&mut self) -> Option<PhysFrame<S>> {
+        let start_address = self.inner.allocate_frame()?.start_address() + self.encryption.bit();
+        Some(PhysFrame::from_start_address(start_address).unwrap())
+    }
+}
+
+/// Page mapper for pages of type <S>
+///
+/// This is equivalent to <x86_64::structures::paging::mapper::Mapper>, but knows about memory
+/// encryption.
+pub trait Mapper<S: PageSize> {
+    unsafe fn map_to_with_table_flags<A>(
+        &mut self,
+        page: Page<S>,
+        frame: PhysFrame<S>,
+        flags: PageTableFlags,
+        parent_table_flags: PageTableFlags,
+        frame_allocator: &mut A,
+    ) -> Result<MapperFlush<S>, MapToError<S>>
+    where
+        A: FrameAllocator<Size4KiB>;
```

Comment:
In here, yes. In general, no.

The mapper needs an allocator for 4K pages in case it needs to allocate new PageTable-s internally, as they are 4K in size.